lib/ash/subject.ex

```diff
@@ -356,4 +431,232 @@ defmodule Ash.Subject do
       %{subject | after_action: subject.after_action ++ [callback]}
     end
   end
+
+  @doc """
+  Adds a before_transaction hook to the subject.
+
+  Before transaction hooks are executed before the database transaction begins.
+  They receive the subject and must return either a modified subject or an error tuple.
+  These hooks are useful for validation, authorization checks, or preparatory logic
+  that should run outside of the transaction.
+
+  ## Parameters
+
+    * `subject` - The subject to add the hook to (Changeset, Query, or ActionInput)
+    * `callback` - Function that takes the subject and returns the subject or `{:error, reason}`
+    * `opts` - Options including `:prepend?` to add at beginning of hooks list
+
+  ## Examples
+
+      # Add validation before transaction starts
+      iex> changeset
+      ...> |> Ash.Subject.before_transaction(fn changeset ->
+      ...>   if valid_state?(changeset) do
+      ...>     changeset
+      ...>   else
+      ...>     {:error, "Invalid state for this operation"}
+      ...>   end
+      ...> end)
+
+      # Add logging for all subject types
+      iex> subject
+      ...> |> Ash.Subject.before_transaction(fn subject ->
+      ...>   Logger.info("Starting transaction for \#{inspect(subject.resource)}")
+      ...>   subject
+      ...> end)
+
+      # Prepend a hook to run first
+      iex> query
+      ...> |> Ash.Subject.before_transaction(check_permissions, prepend?: true)
+
+  ## See also
+
+  - `after_transaction/3` for hooks that run after the transaction completes
+  - `around_transaction/3` for hooks that wrap the entire transaction
+  - `before_action/3` for hooks that run before the action (inside transaction)
+  """
+  @spec before_transaction(
+          subject :: t(),
+          callback :: before_transaction_fun(),
+          opts :: Keyword.t()
+        ) :: t()
+  def before_transaction(subject, callback, opts \\ [])
+
+  def before_transaction(%Ash.Changeset{} = subject, callback, opts) do
+    Ash.Changeset.before_transaction(subject, callback, opts)
+  end
+
+  def before_transaction(%Ash.Query{} = subject, callback, opts) do
+    Ash.Query.before_transaction(subject, callback, opts)
+  end
+
+  def before_transaction(subject, callback, opts) do
+    if opts[:prepend?] do
```

Comment:
We should probably be delegating this to `ActionInput` too. Starting to think we should just be making this a protocol...